Viewing a patient crashes on Galaxy Tab A

Attempting to create or view a patient in v0.12 or v0.13.1 on a Galaxy Tab A immediately crashes the client.

[viewing-patient-crash-logcat.txt](https://github.com/projectbuendia/buendia/files/3358014/viewing-patient-crash-logcat.txt)

Here's the relevant exception:

```
E/AndroidRuntime(13481): FATAL EXCEPTION: main
E/AndroidRuntime(13481): Process: org.projectbuendia.client.dev, PID: 13481
E/AndroidRuntime(13481): java.lang.NoClassDefFoundError: Failed resolution of: Ljava/util/Locale$Category;
E/AndroidRuntime(13481): 	at org.apache.commons.text.ExtendedMessageFormat.<init>(ExtendedMessageFormat.java:146)
E/AndroidRuntime(13481): 	at org.projectbuendia.client.ui.chart.ObsFormat.<init>(ObsFormat.java:97)
E/AndroidRuntime(13481): 	at org.projectbuendia.client.ui.chart.ObsFormat.<init>(ObsFormat.java:115)
E/AndroidRuntime(13481): 	at org.projectbuendia.client.ui.chart.ObsFormat.fromPattern(ObsFormat.java:125)
E/AndroidRuntime(13481): 	at org.projectbuendia.client.models.ChartItem.<init>(ChartItem.java:42)
E/AndroidRuntime(13481): 	at org.projectbuendia.client.sync.ChartDataHelper.getCharts(ChartDataHelper.java:342)
E/AndroidRuntime(13481): 	at org.projectbuendia.client.ui.chart.PatientChartController.<init>(PatientChartController.java:239)
E/AndroidRuntime(13481): 	at org.projectbuendia.client.ui.chart.PatientChartActivity.onCreateImpl(PatientChartActivity.java:252)
E/AndroidRuntime(13481): 	at org.projectbuendia.client.ui.BaseLoggedInActivity.onCreate(BaseLoggedInActivity.java:84)
E/AndroidRuntime(13481): 	at android.app.Activity.performCreate(Activity.java:6609)
E/AndroidRuntime(13481): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1134)
E/AndroidRuntime(13481): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3104)
E/AndroidRuntime(13481): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3261)
E/AndroidRuntime(13481): 	at android.app.ActivityThread.access$1000(ActivityThread.java:219)
E/AndroidRuntime(13481): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1735)
E/AndroidRuntime(13481): 	at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime(13481): 	at android.os.Looper.loop(Looper.java:145)
E/AndroidRuntime(13481): 	at android.app.ActivityThread.main(ActivityThread.java:6939)
E/AndroidRuntime(13481): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(13481): 	at java.lang.reflect.Method.invoke(Method.java:372)
E/AndroidRuntime(13481): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1404)
E/AndroidRuntime(13481): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1199)
E/AndroidRuntime(13481): Caused by: java.lang.ClassNotFoundException: Didn't find class "java.util.Locale$Category" on path: DexPathList[[zip file "/system/framework/android.test.runner.jar", zip file "/data/app/org.projectbuendia.client.dev-1/base.apk"],nativeLibraryDirectories=[/vendor/lib, /system/lib]]
E/AndroidRuntime(13481): 	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
E/AndroidRuntime(13481): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
E/AndroidRuntime(13481): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
E/AndroidRuntime(13481): 	... 22 more
E/AndroidRuntime(13481): 	Suppressed: java.lang.ClassNotFoundException: java.util.Locale$Category
E/AndroidRuntime(13481): 		at java.lang.Class.classForName(Native Method)
E/AndroidRuntime(13481): 		at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
E/AndroidRuntime(13481): 		at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
E/AndroidRuntime(13481): 		at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
E/AndroidRuntime(13481): 		... 23 more
E/AndroidRuntime(13481): 	Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack available
W/ActivityManager(  738):   Force finishing activity 1 org.projectbuendia.client.dev/org.projectbuendia.client.ui.chart.PatientChartActivity
D/InputDispatcher(  738): Focused application set to: xxxx
```

--- a/.gitignore
+++ b/.gitignore
@@ -20,6 +20,7 @@ bin/
 build/
 gen/
 crashlytics-build.properties
+output.json
 
 # Gradle files
 .gradle/
@@ -33,9 +34,6 @@ local.properties
 .idea/copyright/profiles_settings.xml
 .idea/inspectionProfiles/Buendia.xml
 .idea/inspectionProfiles/profiles_settings.xml
-
-
-
 !.idea/inspectionProfiles
 !.idea/copyright
 

--- a/app/src/main/java/org/projectbuendia/client/ui/chart/ObsFormat.java
+++ b/app/src/main/java/org/projectbuendia/client/ui/chart/ObsFormat.java
@@ -94,7 +94,10 @@ public class ObsFormat extends Format {
         }
         mRootObsFormat = Utils.orDefault(rootObsFormat, this);
         try {
-            mFormat = new ExtendedMessageFormat(pattern, new FormatFactoryMap());
+            // It's unsafe to use the ExtendedMessageFormat(pattern, registry) constructor,
+            // as it crashes with a NoClassDefFoundError on java.util.Locale.Category on
+            // Android 5.1.  We must use ExtendedMessageFormat(pattern, locale, registry).
+            mFormat = new ExtendedMessageFormat(pattern, Locale.US, new FormatFactoryMap());
         } catch (IllegalArgumentException e) {
             // Instead of crashing, display the invalid pattern in the output to aid debugging.
             mFormat = new Format() {
@@ -153,9 +156,9 @@ public class ObsFormat extends Format {
         }
 
         /**
-         * ExtendedMessageFormat expects a Map containing FormatFactory instances;
-         * rather than defining a separate FormatFactory class to go with every Format,
-         * we simply return this class, which can instantiate any Format.
+         * ExtendedMessageFormat expects a Map containing FormatFactory instances.
+         * Rather than defining a separate FormatFactory class for every Format,
+         * we return the FormatFactoryMap itself, which can instantiate any Format.
          */
         @Override public @Nullable FormatFactory get(Object name) {
             return FORMAT_CLASSES.containsKey("" + name) ? this : null;

